dataset/map_view.cpp

```diff
@@ -139,7 +139,11 @@ void Dict<Key, Value>::erase(const key_type &key) {
 }
 
 template <class Key, class Value>
-Value Dict<Key, Value>::extract(const key_type &key) {
+Value Dict<Key, Value>::extract(const key_type &key,
+                                const mapped_type *const default_value) {
```

Comment:
The problem with `std::optional` is that it can only store references through `std::reference_wrapper` which makes using it awkward.
Passing by value via `std::optional<mapped_type>` would make a new object (typically `Variable`) every time you use `pop` or copy the object that is passed in. I would prefer reference semantics. But of course I could just overload `pop` and avoid the problem in C++ altogether.

Reply:
I went with overloading in C++. The Python bindings still use pointers but that is invisible to the user.